modules/ml/src/main/java/org/apache/ignite/ml/inference/storage/model/thinclient/ModelStorateThinClientProcessor.java

```diff
@@ -0,0 +1,331 @@
+
+    /**
+     * Reads file from model storage.
+     *
+     * @param reqId Request id.
+     * @param reader Reader.
+     * @return Response.
+     */
+    private ClientResponse readFile(long reqId, BinaryRawReader reader) {
+        String path = reader.readString();
+
+        return modelStorage.lockPaths(() -> {
+            if (!modelStorage.exists(path))
+                return error(reqId, "File not found [path=" + path + "]");
+
+            if(!modelStorage.isFile(path))
+                return error(reqId, "File is not regular file [path" + path + "]");
+
+            return new FileRespose(reqId, modelStorage.getFile(path));
+        }, path);
+    }
+
+    /**
+     * Moves file in model storage.
+     *
+     * @param reqId Request id.
+     * @param reader Reader.
+     * @return Response.
+     */
+    private ClientResponse moveFile(long reqId, BinaryRawReader reader) {
+        String from = reader.readString();
+        String to = reader.readString();
+
+        return modelStorage.lockPaths(() -> {
+            if (!modelStorage.exists(from))
+                return error(reqId, "File not found [path=" + from + "]");
+            if (modelStorage.exists(to))
+                return error(reqId, "File already exists [path=" + to + "]");
+            if (!modelStorage.isFile(from))
+                return error(reqId, "File is not regular file [path=" + from + "]");
+
+            byte[] file = modelStorage.getFile(from);
+            modelStorage.remove(from);
+            modelStorage.putFile(to, file);
+            return new ClientResponse(reqId);
+        }, from, to);
+    }
+
+    /**
+     * Returns statistics about file.
+     *
+     * @param reqId Request id.
+     * @param reader Reader.
+     * @return Response.
+     */
+    private ClientResponse getFileStat(long reqId, BinaryRawReader reader) {
+        String path = reader.readString();
+
+        return modelStorage.lockPaths(() -> {
+            if (!modelStorage.exists(path))
+                return error(reqId, "File not found [path=" + path + "]");
+
+            return new FileStatResponse(reqId, modelStorage.getFileStat(path));
+        }, path);
+    }
+
+    /**
+     * Checks file existance.
+     *
+     * @param reqId Request id.
+     * @param reader Reader.
+     * @return Response.
+     */
+    private ClientResponse isExists(long reqId, BinaryRawReader reader) {
+        String path = reader.readString();
+        return new ClientBooleanResponse(reqId, modelStorage.exists(path));
+    }
+
+    /**
+     * Removes file in model storage.
+     *
+     * @param reqId Request id.
+     * @param reader Reader.
+     * @return Response.
+     */
+    private ClientResponse remove(long reqId, BinaryRawReader reader) {
+        String path = reader.readString();
+
+        return modelStorage.lockPaths(() -> {
+            if (modelStorage.exists(path))
+                modelStorage.remove(path);
+
+            return new ClientResponse(reqId);
```

Comment:
Don't remove recursively in `DefaultModelStorage`.

Reply:
okay